### PR TITLE
fix: make sure `sorter` indices are inbound in `searchsorted`

### LINKED ISSUE #91606: [Crash] `torch.searchsorted` with out-of-bound sorter
### 🐛 Describe the bug

```python
import torch

torch.searchsorted(
    torch.tensor([1, 2, 3]),
    2.5,  # when it becomes <= 2, it won't crash.
    sorter=torch.tensor(
        [0, 1, 2**22]
    ),  # when the last element < 2**21 + 1275831, it may not crash on my machine.
) # crash
```

[`torch.searchsorted`](https://pytorch.org/docs/stable/generated/torch.searchsorted.html) should check the integrity of `sorter` where all elements (which represent indices) should be smaller than `sorted_sequence`'s inner-most dimension size (3 here). Otherwise this is a out-of-bound read which leads to either:
1. crash;
2. silently returned with undefined resutls (which could potientially result in data leakage).

In numpy, the elements of `sorter` will be checked to make sure the data integrity:

```python
import numpy as np

np.searchsorted(
    np.array([1, 2, 3]),
    3.5,  # when it becomes <= 2, it won't crash.
    sorter=np.array(
        [0, 1, 3]
    ),  # 3 is out-of-bound.
)
"""
  File "test.py", line 3, in <module>
    np.searchsorted(
  File "<__array_function__ internals>", line 180, in searchsorted
  File "/miniconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 1387, in searchsorted
    return _wrapfunc(a, 'searchsorted', v, side=side, sorter=sorter)
  File "/miniconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
ValueError: Sorter index out of range.
"""
```

### Versions

<details><summary><b>Env </b> <i>[click to expand]</i></summary>
<div>

```python
"""
Collecting environment information...
PyTorch version: 1.14.0.dev20221202+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 515.86.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] onnx2torch==1.5.3
[pip3] torch==1.14.0.dev20221202+cu117
[pip3] torchaudio==0.14.0.dev20221203+cu117
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.15.0.dev20221203+cpu
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] onnx2torch                1.5.3                    pypi_0    pypi
[conda] torch                     1.14.0.dev20221202+cu117          pypi_0    pypi
[conda] torchaudio                0.14.0.dev20221203+cu117          pypi_0    pypi
[conda] torchtriton               2.0.0+0d7e753227          pypi_0    pypi
[conda] torchvision               0.15.0.dev20221203+cpu          pypi_0    pypi
"""
```

</div>
</details>

### PR DESCRIPTION
**Pull request by ganler**
Fixes #91606

Add a checker to `sorter` to make sure indices are inbound (as NumPy).

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/BucketizationUtils.h [hunk 1]
```diff
@@ -134,6 +134,12 @@ inline void searchsorted_pre_check(
 
     TORCH_CHECK(sorter.scalar_type() == ScalarType::Long, "torch.searchsorted(): sorter must be a tensor of long ",
       "dtype but got dtype ", sorter.scalar_type());
+
+    if (sorter.numel() > 0) {
+      const long innermost_size = sorter.sizes().back();  // get inner most dimension of sorter
+      TORCH_CHECK(sorter.max().item().toLong() < innermost_size, "torch.searchsorted(): sorter index out of range");
+      TORCH_CHECK(sorter.min().item().toLong() >= 0, "torch.searchsorted(): sorter index out of range");
```
Comment (malfet) [inline aten/src/ATen/native/BucketizationUtils.h:141]: Documentation is a bit obscure here, but aren't indices in Pytorch usually wrap around (i.e. `min()` should be `>= -inntermost_size`)

Comment (ganler) [reply aten/src/ATen/native/BucketizationUtils.h:141]: I agree. But the `searchsorted` implementation seems did not well consider such cases.

image

`0` and `-1` in sorter should represent the same indices for a single-elem vector but the results are different. Not sure if we wanted to fix it and allow such usage, or just disable it for now.

Comment (ngimel) [reply aten/src/ATen/native/BucketizationUtils.h:141]: nit: use `aminmax` for slightly better perf.

**Comment by ganler** [comment]
It seems one of the task somehow gets stuck https://github.com/pytorch/pytorch/actions/runs/4188232059/jobs/7260592907

**Pull request closed without merging**